src/libertem/udf/base.py

```diff
@@ -829,106 +881,116 @@ def _get_dtype(self, dtype):
 
     def run_for_partition(self, partition: Partition, roi):
         with set_num_threads(1):
-            backend = get_backend()
-            dtype = self._get_dtype(partition.dtype)
-            meta = UDFMeta(
-                partition_shape=partition.slice.adjust_for_roi(roi).shape,
-                dataset_shape=partition.meta.shape,
-                roi=roi,
-                dataset_dtype=partition.dtype,
-                input_dtype=dtype,
-                tiling_scheme=None,
-                backend=backend,
-            )
-            udfs = self._udfs
-            for udf in udfs:
-                udf.set_backend(backend)
-                udf.set_meta(meta)
-                udf.init_result_buffers()
-                udf.allocate_for_part(partition, roi)
-                udf.init_task_data()
-                if hasattr(udf, 'preprocess'):
-                    udf.clear_views()
-                    udf.preprocess()
-            neg = Negotiator()
-            # FIXME take compute backend into consideration as well
-            # Other boundary conditions when moving input data to device
-            tiling_scheme = neg.get_scheme(
-                udfs=udfs,
-                partition=partition,
-                read_dtype=dtype,
-                roi=roi,
-            )
-
-            # FIXME: don't fully re-create?
-            meta = UDFMeta(
-                partition_shape=partition.slice.adjust_for_roi(roi).shape,
-                dataset_shape=partition.meta.shape,
-                roi=roi,
-                dataset_dtype=partition.dtype,
-                input_dtype=dtype,
-                tiling_scheme=tiling_scheme,
-                backend=backend,
-            )
-            for udf in udfs:
-                udf.set_meta(meta)
-            # print("UDF TilingScheme: %r" % tiling_scheme.shape)
-
-            # FIXME pass information on target location (numpy or cupy)
-            # to dataset so that is can already move it there.
-            # In the future, it might even decode data on the device instead of CPU
-            tiles = partition.get_tiles(
-                tiling_scheme=tiling_scheme,
-                roi=roi, dest_dtype=dtype,
-            )
-            # Guaranteed to be the same by UDFTask.get_resources()
-            xp = udfs[0].xp
-            for tile in tiles:
-                # Work-around, should come from dataset later
-                device_tile = xp.asanyarray(tile)
+            try:
+                previous_id = False
```

Comment:
Hmm, this is starting to reach indentation levels that are a bit too much - maybe the `with` and `try...finally` blocks can be done in this function, and the content can be moved into its own method? As this method is central to UDF processing, we need to take care to keep it readable, and over time try to increase readability, not the other way around :grin:

Reply:
Ok, I've split it up, it was about time...